### PR TITLE
Added C# implementation of System.Math.ScaleB and System.MathF.ScaleB…

### PR DESCRIPTION
**Pull request by alexcovington**
Implemented System.Math.ScaleB and System.MathF.ScaleB in managed C#, added test cases.

Submitting as a draft pull request first because I'm not quite sure where/how to integrate the tests I've written. System.Private.CoreLib does not have a test directory and I'm having trouble getting `build.cmd` to find the tests. Any feedback would be appreciated!

This is to address [this issue](https://github.com/dotnet/runtime/issues/40956) regarding a performance difference in `System.Math.ScaleB` and `System.MathF.ScaleB` on Windows and Linux. This PR improves the speed of the function on both platforms and removes the previous bindings/unmanaged implementations.

Edit: updated description and added link to issue.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42476) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42476) before we can accept your contribution.<br/><br/>:x: alexcovington [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42476)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=42476) it.</sub>

**Comment by am11** [comment]
Existing tests are based on xunit framework and located at `src/libraries/System.Runtime.Extensions/tests/System/Math{F}.cs`. 

From PowerShell, you can do:

```powershell
# build the whole thing just once
$ cd runtime
$ ./build.cmd -c Release

# then switch to the 'test' directory of library of interest, in this case
$ cd src/libraries/System.Runtime.Extensions/tests/

# build and run tests using restored toolset
$ ../../../../.dotnet/dotnet build /t:Test -c Release
# at this point, can modify code in src/libraries/System.Private.CoreLib/src/System and just repeat the command
# on line above to rebuild src+test (and their dependencies) for this library and then run tests.
```

File: src/coreclr/src/System.Private.CoreLib/src/System/Math.CoreCLR.cs
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/src/System/Math.CoreCLR.cs:101]: Could you please delete it from the unmanaged runtime:

src/coreclr/src/classlibnative/float/floatdouble.cpp
src/coreclr/src/classlibnative/float/floatsingle.cpp
src/coreclr/src/classlibnative/inc/floatdouble.h
src/coreclr/src/classlibnative/inc/floatsingle.h
src/coreclr/src/vm/ecalllist.h
Everything that references scalbn and scalbnf under src/coreclr/src/pal

Comment (am11) [reply src/coreclr/src/System.Private.CoreLib/src/System/Math.CoreCLR.cs:101]: Also, I think these lines can be deleted from mono: https://github.com/dotnet/runtime/blob/2a1735b162437bb35a112014092a834c2496c3e3/src/mono/mono/metadata/sysmath.c#L221-L225 https://github.com/dotnet/runtime/blob/2a1735b162437bb35a112014092a834c2496c3e3/src/mono/mono/metadata/sysmath.c#L390-L394 https://github.com/dotnet/runtime/blob/2a1735b162437bb35a112014092a834c2496c3e3/src/mono/mono/metadata/icall-def-netcore.h#L148 https://github.com/dotnet/runtime/blob/2a1735b162437bb35a112014092a834c2496c3e3/src/mono/mono/metadata/icall-def-netcore.h#L177 https://github.com/dotnet/runtime/blob/2a1735b162437bb35a112014092a834c2496c3e3/src/mono/mono/metadata/icall-decl.h#L142 https://github.com/dotnet/runtime/blob/2a1735b162437bb35a112014092a834c2496c3e3/src/mono/mono/metadata/icall-decl.h#L146

Comment (EgorBo) [reply src/coreclr/src/System.Private.CoreLib/src/System/Math.CoreCLR.cs:101]: @am11 `icall-decl.h` should be preserved, it's used by 6.x mono

Comment (am11) [reply src/coreclr/src/System.Private.CoreLib/src/System/Math.CoreCLR.cs:101]: Ah, ok I see. We can then just delete two lines in `icall-def-netcore.h` and leave `icall-decl.h` and `sysmath.c`. Updated my comment above.

File: src/libraries/System.Private.CoreLib/tests/math/MathFTests.cs
Comment (hughbe) [inline src/libraries/System.Private.CoreLib/tests/math/MathFTests.cs:10]: Why are we not using xunit tests in System.Runtime.Tests?

Comment (hughbe) [reply src/libraries/System.Private.CoreLib/tests/math/MathFTests.cs:10]: Saw your comment above - take a look at https://github.com/dotnet/runtime/blob/master/src/libraries/System.Runtime.Extensions/tests/System/MathF.cs

Comment (alexcovington) [reply src/libraries/System.Private.CoreLib/tests/math/MathFTests.cs:10]: This is exactly what I was looking for before! I just missed it when I was looking for current tests. I'll just add my new test cases to the existing `ScaleB` tests.

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:1106]: nit: use the [documented](https://docs.microsoft.com/en-us/dotnet/api/system.math.scaleb?view=netcore-3.1) parameter names:
```suggestion
        public static double ScaleB(double x, int n)
```

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:1118]: style nit: surround with brackets to match the existing style - elsewhere in this file, curly brackets around single line conditions are used https://github.com/dotnet/runtime/blob/57c49318b4de61a17cce116c3d9a2dee3cd4f31a/src/libraries/System.Private.CoreLib/src/System/Math.cs#L301-L318

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:1133]: style nit: add an empty line between 1132 and 1133.

**Comment by vargaz** [comment]
The mono changes look ok to me.

**Comment by CoffeeFlux** [comment]
For Mono, please wrap the corresponding icall implementations in `#ifndef ENABLE_NETCORE`. We want them around for framework Mono but not netcore. https://github.com/dotnet/runtime/blob/1905bde27967bff71c95ec5cbf2e099d80dce5f4/src/mono/mono/metadata/sysmath.c#L222

File: src/coreclr/src/pal/inc/pal.h
Comment (jkotas) [inline src/coreclr/src/pal/inc/pal.h:3858]: Could you please also delete `src\coreclr\src\pal\tests\palsuite\c_runtime\scalbn` and ``src\coreclr\src\pal\tests\palsuite\c_runtime\scalbnf`

Comment (am11) [reply src/coreclr/src/pal/inc/pal.h:3858]: The output of `git grep scalbn -- :/src/coreclr/src/pal` also shows related entries in:

* `src/coreclr/src/pal/tests/palsuite/CMakeLists.txt`
* `src/coreclr/src/pal/tests/palsuite/compilableTests.txt`
* `src/coreclr/src/pal/tests/palsuite/paltestlist.txt`

File: src/mono/mono/metadata/sysmath.c
Comment (am11) [inline src/mono/mono/metadata/sysmath.c:232]: nit: mono's C code uses tabs for indentation:
```suggestion
	return scalbn(x, n);
```
(same on line 403)

**Comment by CoffeeFlux** [comment]
I apologize, it looks like these are netcore-only; I'd thought otherwise based on comments above. In that case, we can remove the entries in icall-decl.h and sysmath.c entirely for Mono. Again, sorry for the churn on that, totally my fault.

**Comment by am11** [comment]
@CoffeeFlux, good to know, thanks. Would it make sense to move netcore-specific declarations to `icall-def-netcore.h` (in a separate PR)?

**Comment by CoffeeFlux** [comment]
It's a little complicated.

We have all the netcore definitions in `icall-def-netcore.h`. Some of the icalls also have identical versions in framework and so we share the implementation. Some icalls use handles, and for which we are generating the declaration inside the macro, but that comes at a performance cost. For others, we don't use the handles (hence the `NOHANDLES` macro) and have a separate definition, which we put in the `icall-decl.h` file. I think the current state of affairs is okay, it's just my bad on assuming things and not checking to make sure the declaration wasn't already wrapped in an `#if ENABLE_NETCORE`.

**Comment by alexcovington** [comment]
@am11 Thanks! I have to step away for a little bit, but I'll resolve it later this afternoon.

**Comment by CoffeeFlux** [comment]
Looks good now from a Mono perspective, thanks!

**Comment by danmosemsft** [comment]
What's next steps here?
cc @tannergooding

**Comment by tannergooding** [comment]
> What's next steps here?

I'll get this reviewed either today or tomorrow. I'm not aware of anything else that would be blocking.

It would likely be good to get perf numbers for x64, x86, ARM32, and ARM64. @DrewScoggins, is there currently an easy way to get these for a particular benchmark (aside from manually running them)?

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:1106]: Was this ported from an existing implementation?

Based on https://github.com/dotnet/runtime/issues/40956#issuecomment-693022684, it sounds like its based on the MUSL implementation which is MIT (https://git.musl-libc.org/cgit/musl/tree/COPYRIGHT) and so we need a comment around here and also need to update the https://github.com/dotnet/runtime/blob/master/THIRD-PARTY-NOTICES.TXT.

CC. @richlander on if there is anything else required.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:1106]: The MUSL implementation came from Sun's implementation, so we may need the notice for that too (or just have notice for the Sun's implementation if it was not changed for MUSL).

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:1109]: Its probably not particularly important here, but the MSVC and the JIT tends to emit (assuming no PGO data):
```
if (condition) {
  // predicted
}
else {
  // unpredicted
}
```

This is inverse from what Clang/GCC do, so the `n > 1023` and `n < -1022` cases are going to be predicted as taken by default, while I believe they are edge cases in practice.

You can compare JIT (noting this is 3.1 output, not .NET 5, so there may be a couple minor differences): <https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgNIAO1wYGAEwgBXYABtxAZRYBBADIBRAEIB9FuQYBeBgA4AdAE4rVlEgCsSAOwo65ci6t6AajoABncAbgYAekiGYIRyXnJg0jouHn5hUXEpGXklVQ1tfWMWUgsGUhtSUhdQuis/WvdSYPIUPTgQqwjo2PjeOCTqtO4MkTEJCGlZBWU1Bk1dQxNGSztg0PI7OxqUT2Dt0h6YuISXVOoRpkZiciQ8uY0wbDUDAApZgoYENAYBSVkkgAlFdONQeBCHl8AJ7lBBhK4QgQAMwYb0kDAAfAwhnQQeDIYSwYTCbCAFSWRbFEzkBEEkk8DFwSy4ukMpGo9FYnHJPGIknE9mkhgUhZFZamNlCyFMlm8qXSngotEY7G4/GKiGCzUy8qs/lCgC+BshxvpppNMBUuHEyq5AB4GINkqQNQKTRDyZTxSUymSxUsSnQFeyMQE5dUnQxziGGXaMY7ndU3ULtZqvQHqaURZmJcGPSSwxGynBo/nzYr4wxE0NXQWifXQ+Uk0dGxCzYqOySzSbPvNFOUDAIMCwpkIYFAxFAbABJAGoIcYXAAFQgABF8mo3tuVFMAOZAt5xOjI1EBBjA6uOlyuoGxnjEdwMckMRQhs2GoA>

To MSVC and Clang: <https://godbolt.org/z/MEYGcz>

File: src/libraries/System.Runtime.Extensions/tests/System/MathF.cs
Comment (tannergooding) [inline src/libraries/System.Runtime.Extensions/tests/System/MathF.cs:200]: It looks like the diff for this file is bigger because the whitespace was stripped from the attributes.

It's probably a losing battle to try and fight with the formatter to keep the "columns" aligned, so I don't think it needs to be fixed; just wanted to call it out.

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:1106]: It might be good to add a small comment explaining how this function works.

Just something similar to or roughly detailing that `scalbn(x, n)` is `x * 2^n`. So we construct a `double` from `2^n` by building the correct biased exponent (https://github.com/dotnet/runtime/pull/42476/files#diff-29944976f2a0c621a8f96a2e75ccdf55R1138). 

If `n` is greater than the maximum exponent (`1023` for double, `127` for float) or smaller then the minimum exponent (`-1022` for double, `-126` for float), then we need to make an adjustment to compute the correct result.

**Comment by DrewScoggins** [comment]
> > What's next steps here?
> 
> I'll get this reviewed either today or tomorrow. I'm not aware of anything else that would be blocking.
> 
> It would likely be good to get perf numbers for x64, x86, ARM32, and ARM64. @DrewScoggins, is there currently an easy way to get these for a particular benchmark (aside from manually running them)?

We already generate reports for the category subsets that are not Library and Runtime, but looking at the performance tests for this space, they are not in their own category. So what we can do is just generate a report that covers just the performance tests that we care about by name. We won't have this data for ARM32, as we don't have that hardware in the perf lab, but we should be able to generate the data for all of the other configs.

**Comment by tannergooding** [comment]
> Also this would be after the change already went in.

Right, I know (roughly) how to do that bit. I was interested if we had any way to kick of a perf run against the current PR and to pull that info out before merging, short of just checking it out and running the tests locally on my own hardware.

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (alexcovington) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:1109]: This is good to know. I tried reworking the logic to take this into consideration, roughly something like:

```
if (n <= 127 && n >= -126) {
    // n is in expected range, just compute and return result
} else {
    if (n > 127) {
        // Make adjustments
    }
    else if (n < -126) {
        // Make adjustments
    }
}
```

But I'm not seeing an improvement in perf. I can push if you'd like to take a look, but not sure it's worth it.

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (alexcovington) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:1106]: Yes, this was based on the musl implementation. I added links to the source for each function and included both the musl and Sun license. Let me know if I missed or can add anything.

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:1115]: > I can push if you'd like to take a look, but not sure it's worth it.

Right, I don't think in this particular case it is particular import, I just wanted to call it out as it might be relevant if additional work is done in this or related areas later.

**Comment by ViktorHofer** [comment]
// Auto-generated message

https://github.com/dotnet/runtime/commit/69e114c1abf91241a0eeecf1ecceab4711b8aa62 which was merged 12/7 removed the intermediate src/coreclr/**src**/ folder. This PR needs to be updated as it touches files in that directory which causes conflicts.

To update your commits you can use this bash script: https://gist.github.com/ViktorHofer/6d24f62abdcddb518b4966ead5ef3783. Feel free to use the comment section of the gist to improve the script for others.

**Comment by tannergooding** [comment]
Is this currently blocked on anything and if so, can we help push it forward?

**Comment by alexcovington** [comment]
@tannergooding My apologies, I think this slipped through the cracks on my end. I'll merge with master and push again.

**Comment by tannergooding** [comment]
Thanks, and no worries! Just wanted to check in and see (and make sure I wasn't missing some action item on my end) 😄

File: src/libraries/System.Private.CoreLib/src/System/Math.cs
Comment (danmosemsft) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:1409]: I was thinking that if those edge cases were moved into their own method, this could become inlineable. But it doesn't seem to get inlined even then.

File: src/libraries/System.Private.CoreLib/src/System/MathF.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/MathF.cs:481]: Just noting, since this PR was opened, AMD open sourced their own `libm` implementation which includes `scaleb`: https://github.com/amd/aocl-libm-ose/blob/7193ff6fdf72ecd68df4e0e19262da84f1d95a3e/src/ref/scalbnf.c

It notably has a different algorithm and error handling logic (unnecessary in the C# implementation) but it might be worth comparing in the future to see which is better.

**Review by tannergooding — APPROVED** [review]
LGTM. This should be good to merge assuming @jkotas doesn't have any other feedback about the JIT side changes.

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!

**Comment by tannergooding** [comment]
Thanks again for the contribution @alexcovington!

@adamsitnik, @kunalspathak, @DrewScoggins we should validate this shows up as an improvement in perf triage.

File: src/mono/mono/metadata/sysmath.c
Comment (am11) [inline src/mono/mono/metadata/sysmath.c:210]: @vargaz, I have noticed that in some PRs, instead of deleting the now-unused icall, the function is wrapped in `#ifndef ENABLE_NETCORE`. Should we do the same here for https://github.com/mono/mono/pull/20424?

Comment (CoffeeFlux) [reply src/mono/mono/metadata/sysmath.c:210]: If the icall is used for framework Mono we want to ifdef it out, as the code is mirrored to `mono/mono` and still required there. In this case the icall was netcore-specific, so getting rid of it entirely is preferable.

On a related note, this was a nice reminder to merge the corresponding mono/mono PR :)

**Comment by DrewScoggins** [comment]
Here is the improvement on Ubuntu

https://github.com/DrewScoggins/performance-2/issues/3838

Windows x64

https://github.com/DrewScoggins/performance-2/issues/3857

Arm64

https://github.com/DrewScoggins/performance-2/issues/3848

Windows x86

https://github.com/DrewScoggins/performance-2/issues/3867

**Pull request merged**
